service/history/workflow/mutable_state_impl.go

```diff
@@ -4825,12 +4825,42 @@ func (ms *MutableStateImpl) RetryActivity(
 	if err := ms.checkMutability(opTag); err != nil {
 		return enumspb.RETRY_STATE_INTERNAL_SERVER_ERROR, err
 	}
-	activityVisitor := newActivityVisitor(ai, failure, ms.timeSource)
-	if state := activityVisitor.State(); state != enumspb.RETRY_STATE_IN_PROGRESS {
-		return state, nil
+	if !ai.HasRetryPolicy {
+		return enumspb.RETRY_STATE_RETRY_POLICY_NOT_SET, nil
+	}
+	if ai.CancelRequested {
+		return enumspb.RETRY_STATE_CANCEL_REQUESTED, nil
+	}
+
+	if !isRetryable(failure, ai.RetryNonRetryableErrorTypes) {
+		return enumspb.RETRY_STATE_NON_RETRYABLE_FAILURE, nil
+	}
+
+	retryMaxInterval := ai.RetryMaximumInterval
+	// if a delay is specified by the application it should override the maximum interval set by the retry policy.
+	delay := nextRetryDelayFrom(failure)
+	if delay != nil {
+		retryMaxInterval = durationpb.New(*delay)
 	}
+
+	now := ms.timeSource.Now().In(time.UTC)
+	retryBackoff, retryState := nextBackoffInterval(
+		ms.timeSource.Now().In(time.UTC),
+		ai.Attempt,
+		ai.RetryMaximumAttempts,
+		ai.RetryInitialInterval,
+		retryMaxInterval,
+		ai.RetryExpirationTime,
+		ai.RetryBackoffCoefficient,
+		makeBackoffAlgorithm(delay),
+	)
+	if retryState != enumspb.RETRY_STATE_IN_PROGRESS {
+		return retryState, nil
+	}
+
+	nextScheduledTime := now.Add(retryBackoff)
 	nextAttempt := ai.Attempt + 1
-	if err := ms.taskGenerator.GenerateActivityRetryTasks(ai.ScheduledEventId, activityVisitor.NextScheduledTime(), nextAttempt); err != nil {
+	if err := ms.taskGenerator.GenerateActivityRetryTasks(ai.ScheduledEventId, nextScheduledTime, nextAttempt); err != nil {
```

Comment:
nit: I'd move this logic after updateActivityInfoForRetries and revert the changes previously made to GenerateActivityRetryTasks, so that GenerateActivityRetryTasks still takes in activityInfo and generates task based on that (instead of assuming things like activity version == mutable state current version).

Reply:
I made the changes. But there was a test case specifically checking that the sizes should not be updated when we cannot generate retry activity task. So I'm updating the sizes after the task has been generated. Please take another quick look.